### PR TITLE
Yuval rio pid

### PR DESCRIPTION
**Pull request by YuvalLevy1**

### CODE REVIEW

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    private Notifier notifier;
+
+    /**
+     * The PIDController calculates the wanted speed to reach target.
+     */
+    private PIDController controller;
+    /**
+     * The proportional component of the loop.
+     */
+    private Supplier<Double> kp;
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: rename `kp` to `kP` etc

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    /**
+     * Returns my current location.
+     */
+    private Supplier<Double> PIDSource;
+
+    /**
+     * The calculation returned from the pid.
+     */
+    private Supplier<Double> PIDCalculations;
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: replace with a `Consumer<Double>` named `output`, this consumer should receive the result of the PID calculations done in the notifier thread

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    public RioPIDLoop(double kp, double ki, double kd, double tolerance,
+                      double setpoint, Supplier<Double> PIDSource, Consumer<Double> speedConsumer) {
+        this(() -> kp, () -> ki, () -> kd, () -> tolerance, () -> setpoint, PIDSource, speedConsumer, () -> 0.0, Frequency.DEFAULT);
+    }
+
+    @Override
+    public void enable() {
+        controller = new PIDController(kp.get(), ki.get(), kd.get(), frequency.period);
+        notifier = new Notifier(this::periodic);
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: declare this in the constructor

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    private void periodic() {
+        PIDCalculations = () -> controller.calculate(PIDSource.get(), setpoint.get());
+    }
+
+    @Override
+    public void disable() {
+        notifier.stop();
+        notifier.close();
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: do not `close` the notifier here

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+
+    public void enableContinuousInput(double minimumValue, double maximumValue) {
+        this.controller.enableContinuousInput(minimumValue, maximumValue);
+    }
+
+    @Override
+    public void update() {
+        speedConsumer.accept(PIDCalculations.get());
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: replace current `update` method with a method that uses `controller.setSetpoint` and is synchronized on `controller`

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    @Override
+    public void enable() {
+        controller = new PIDController(kp.get(), ki.get(), kd.get(), frequency.period);
+        notifier = new Notifier(this::periodic);
+        notifier.startPeriodic(frequency.period);
+    }
+
+    private void periodic() {
+        PIDCalculations = () -> controller.calculate(PIDSource.get(), setpoint.get());
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: replace current `periodic` method with a method that uses `controller.calculate`, runs the `PIDOutput` with the produced output and is synchronized on `controller`

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    /**
+     * The frequency the PID loop runs in.
+     */
+    private Frequency frequency;
+
+    /**
+     * Returns my current location.
+     */
+    private Supplier<Double> PIDSource;
```
Comment (Ergoold) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: rename this to `source`

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 8]
```diff
@@ -92,20 +87,21 @@
     /**
      * The output of RioPIDLoop.
      */
-    private Consumer<Double> speedConsumer;
+    private Consumer<Double> output;
 
     public RioPIDLoop(Supplier<Double> kp, Supplier<Double> ki, Supplier<Double> kd, Supplier<Double> tolerance,
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: please rename `PIDSource` to `source`,  and `speedConsumer` to `output`

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 9]
```diff
@@ -145,34 +141,40 @@ public RioPIDLoop(double kp, double ki, double kd, double tolerance,
...
-        controller = new PIDController(kp.get(), ki.get(), kd.get(), frequency.period);
-        notifier = new Notifier(this::periodic);
+        controller = new PIDController(kP.get(), kI.get(), kD.get(), frequency.period);
         notifier.startPeriodic(frequency.period);
     }
 
     private void periodic() {
-        PIDCalculations = () -> controller.calculate(PIDSource.get(), setpoint.get());
+        synchronized (controller) {
+            output.accept(controller.calculate(source.get(), setpoint.get()));
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: remove the argument `setpoint.get()`

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 10]
```diff
@@ -147,7 +147,7 @@ public void enable() {
 
     private void periodic() {
         synchronized (controller) {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: please replace this `syncronized` block with a use of a `ReetrantLock` you can read about it [here](https://www.geeksforgeeks.org/reentrant-lock-java/).
you will need to store a private instance of `ReentrantLock`.

**Pull request merged**